File: ruby/test/protocol_v1_json_test.rb
# frozen_string_literal: true

require "test_helper"

class ProtocolV1JSONTest < Minitest::Test
  Command = ActionCableClient::Protocol::Command

  # What a decoded frame carries when the frame said nothing about it.
  NOTHING = { kind: nil, identifier: nil, message: "", reason: nil, reconnect: false }.freeze

  def setup
    @protocol = ActionCableClient::Protocol::V1JSON.new
  end

  def test_v1_json_subprotocol
    assert_equal "actioncable-v1-json", @protocol.subprotocol
  end

  def test_v1_json_encode
    {
      Command.new(name: :subscribe, identifier: %({"channel":"RoomChannel"})) =>
        %({"command":"subscribe","identifier":"{\\"channel\\":\\"RoomChannel\\"}"}),
      Command.new(name: :unsubscribe, identifier: %({"channel":"RoomChannel"})) =>
        %({"command":"unsubscribe","identifier":"{\\"channel\\":\\"RoomChannel\\"}"}),
      Command.new(name: :message, identifier: %({"channel":"RoomChannel"}), data: %({"action":"speak"})) =>
        %({"command":"message","identifier":"{\\"channel\\":\\"RoomChannel\\"}","data":"{\\"action\\":\\"speak\\"}"})
    }.each do |command, encoded|
      assert_equal encoded, @protocol.encode(command)
    end
  end

  def test_v1_json_decode
    [
      [ %({"type":"welcome"}), { kind: :welcome } ],
      [ %({"type":"ping","message":1755400000}), { kind: :ping, message: "1755400000" } ],
      [ %({"type":"disconnect","reason":"server_restart","reconnect":true}),
        { kind: :disconnect, reason: "server_restart", reconnect: true } ],
      [ %({"type":"confirm_subscription","identifier":"{\\"channel\\":\\"RoomChannel\\"}"}),
        { kind: :confirmation, identifier: %({"channel":"RoomChannel"}) } ],
      [ %({"type":"reject_subscription","identifier":"{\\"channel\\":\\"RoomChannel\\"}"}),
        { kind: :rejection, identifier: %({"channel":"RoomChannel"}) } ],
      [ %({"identifier":"{\\"channel\\":\\"RoomChannel\\"}","message":{"body":"Hello!"}}),
        { kind: :message, identifier: %({"channel":"RoomChannel"}), message: %({"body":"Hello!"}) } ],
      [ %({"type":"something_new","identifier":"x","message":"anything"}),
        { kind: :message, identifier: "x", message: %("anything") } ]
    ].each do |payload, expected|
      incoming = @protocol.decode(payload)
      decoded = { kind: incoming.kind, identifier: incoming.identifier, message: incoming.message.to_s,
                  reason: incoming.reason, reconnect: incoming.reconnect? }

      assert_equal NOTHING.merge(expected), decoded, payload
    end
  end

  def test_v1_json_decode_garbage
    assert_raises(ActionCableClient::Error) { @protocol.decode("not json") }
  end
end

File: typescript/tests/protocol-v1-json.test.ts
import { describe, expect, it } from "vitest";
import { DisconnectReason, V1JSON, type Command, type IncomingKind } from "../src/index.js";

/** An expected frame, with the message as the text it decodes to. */
interface Decoded {
  kind: IncomingKind;
  identifier?: string;
  message?: string;
  reason?: string;
  reconnect?: boolean;
}

describe("V1JSON", () => {
  it("names its subprotocol", () => {
    expect(new V1JSON().subprotocol).toBe("actioncable-v1-json");
  });

  it("encodes", () => {
    const commands: Array<[Command, string]> = [
      [
        { name: "subscribe", identifier: `{"channel":"RoomChannel"}` },
        `{"command":"subscribe","identifier":"{\\"channel\\":\\"RoomChannel\\"}"}`,
      ],
      [
        { name: "unsubscribe", identifier: `{"channel":"RoomChannel"}` },
        `{"command":"unsubscribe","identifier":"{\\"channel\\":\\"RoomChannel\\"}"}`,
      ],
      [
        {
          name: "message",
          identifier: `{"channel":"RoomChannel"}`,
          data: `{"action":"speak"}`,
        },
        `{"command":"message","identifier":"{\\"channel\\":\\"RoomChannel\\"}","data":"{\\"action\\":\\"speak\\"}"}`,
      ],
    ];

    for (const [command, encoded] of commands) {
      expect(new V1JSON().encode(command), command.name).toBe(encoded);
    }
  });

  it("decodes", () => {
    const frames: Array<[string, Decoded]> = [
      [`{"type":"welcome"}`, { kind: "welcome" }],
      [`{"type":"ping","message":1755400000}`, { kind: "ping", message: "1755400000" }],
      [
        `{"type":"disconnect","reason":"server_restart","reconnect":true}`,
        { kind: "disconnect", reason: DisconnectReason.serverRestart, reconnect: true },
      ],
      [
        `{"type":"confirm_subscription","identifier":"{\\"channel\\":\\"RoomChannel\\"}"}`,
        { kind: "confirmation", identifier: `{"channel":"RoomChannel"}` },
      ],
      [
        `{"type":"reject_subscription","identifier":"{\\"channel\\":\\"RoomChannel\\"}"}`,
        { kind: "rejection", identifier: `{"channel":"RoomChannel"}` },
      ],
      [
        `{"identifier":"{\\"channel\\":\\"RoomChannel\\"}","message":{"body":"Hello!"}}`,
        {
          kind: "message",
          identifier: `{"channel":"RoomChannel"}`,
          message: `{"body":"Hello!"}`,
        },
      ],
      [
        `{"type":"something_new","identifier":"x","message":"anything"}`,
        { kind: "message", identifier: "x", message: `"anything"` },
      ],
    ];

    for (const [payload, expected] of frames) {
      const incoming = new V1JSON().decode(payload);

      expect(incoming.kind, payload).toBe(expected.kind);
      expect(incoming.identifier, payload).toBe(expected.identifier ?? "");
      expect(incoming.message.toString(), payload).toBe(expected.message ?? "");
      expect(incoming.reason, payload).toBe(expected.reason ?? "");
      expect(incoming.reconnect, payload).toBe(expected.reconnect ?? false);
    }
  });

  it("refuses garbage", () => {
    expect(() => new V1JSON().decode("not json")).toThrow(SyntaxError);
    expect(() => new V1JSON().decode(`["not an object"]`)).toThrow(SyntaxError);
  });
});
